Adding USERPASS Auth option to Socks reproxy

nirvantyagi opened the pull request:
Adds USERPASS Auth as an option for authentication to Socks reproxy server. Username is pulled from rtcToNet's userId field. This means that for Tor, each Getter will have a different username leading to a different route through the Tor network. Password is empty.

Considerations
- The username/password is set inside rtc-to-net.ts and is hard-coded in. It does not currently support a Getter/Sharer inputting a specific username/password to use for a specific Socks reproxy. In this sense, even though the rest of the code is modular for ANY Socks reproxy, this is specific to a Tor-like socks server where the actual username/password credentials do not matter - they are only used for identification.

- In generic_ui/scripts/model.ts, the default values for auth type are set. However, I was unable to import the enum values from lib/socks/headers.ts due to a build error in browserfy/vulcanize. @trevj  Currently, the enum values are just hard-coded in. (0 = NOAUTH, 2 = USERPASS).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2637)
<!-- Reviewable:end -->

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/generic_core/globals.ts, line 46 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6wM4TVvfHvUdhW9e:-KOL6wM4TVvfHvUdhW9f:2090103722) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/globals.ts#L46)):*
> ```XML
>     enabled: false,
>     socksEndpoint: {address: '127.0.0.1', port: 9050},
>     auth: [socks.Auth.NOAUTH]
> ```

I find this very strange.  What is being set globally by this?

---

*[src/generic_ui/scripts/model.ts, line 78 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6k8pzaYvchpatRo7:-KOL6k8pzaYvchpatRo8:-121209743) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_ui/scripts/model.ts#L78)):*
> ```XML
>       enabled: false,
>       socksEndpoint: {address: '127.0.0.1', port: 9050},
>       auth: [0]
> ```

Will this `auth` parameter show up in Advanced Settings?

---

*[src/interfaces/uproxy_core_api.ts, line 94 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL7ASL-txZeD77HvuI:-KOL7ASMtsp18rX3L2Cb:-1617953633) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/interfaces/uproxy_core_api.ts#L94)):*
> ```XML
>   enabled       :boolean;      // Reproxy through socks is enabled
>   socksEndpoint :net.Endpoint; // Endpoint through which to reproxy
>   auth          :socks.Auth[];   // Type of socks auth to attempt
> ```

I don't understand what this is for.

---

*[src/lib/socks/headers.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5x3PGGr63EZ1H8em:-KOL5x3PGGr63EZ1H8en:-818828040) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L10)):*
> ```XML
> import net = require('../net/net.types');
> 
> // VERSION - Protocol versions
> ```

Hmm.  I'm not sure how I feel about this.  There's something odd about mixing version numbers of two different versioned things in one enum.

---

*[src/lib/socks/headers.ts, line 239 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL4mBVQZmPd1jAUTIS:-KOL4mBVQZmPd1jAUTIT:-675432164) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L239)):*
> ```XML
> //            +----+------+----------+------+----------+
> export function composeUserPassRequest(creds:UserPass) : ArrayBuffer {
>   var ulen :number = 2*creds.username.length; // 2 bytes per char
> ```

Nit: spaces around `*`

---

*[src/lib/socks/headers.ts, line 275 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5C7tBDrisOF00LRo:-KOL5C7tBDrisOF00LRp:-893515543) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L275)):*
> ```XML
> //                         | 1  |   1    |
> //                         +----+--------+
> export function composeUserPassResponse(success:boolean) : ArrayBuffer {
> ```

Please add unit tests for these four new functions to headers.spec.ts.

---

*[src/lib/socks/headers.ts, line 277 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6_z3b0DDNBi4yo1y:-KOL6_z4y-h17btQwAAt:-166993983) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L277)):*
> ```XML
> export function composeUserPassResponse(success:boolean) : ArrayBuffer {
>   var responseBytes :Uint8Array = new Uint8Array(2);
>   responseBytes[0] = Version.VERSION1; // Only support for version 1 of UserPass protocol.
> ```

Nit: two spaces between `;` and `//`

---

*[src/lib/socks/headers.ts, line 572 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5mGBCBp0Z7rFqOKh:-KOL5mGBCBp0Z7rFqOKi:1408335273) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L572)):*
> ```XML
> 
> // string to ArrayBuffer
> function str2ab(str:string) : ArrayBuffer {
> ```

I think these functions are already available from lib/arraybuffers/arraybuffers.ts.  You might as well just import that instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

dborkan commented:
*[src/generic_core/remote-connection.ts, line 157 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOLHEj40e8ca40brp3r:-KOLHEj40e8ca40brp3s:-1135239412) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/remote-connection.ts#L157)):*
> ```XML
>         allowNonUnicast: globals.settings.allowNonUnicast,
>         reproxy: globals.settings.reproxy,
>         userId: this.userId_
> ```

This is kind of a nitpick given that multiple instance support may be going away with HotPot, but I think this should probably be the remote instance id, not the user id.  uProxy currently supports sharing with multiple devices (instances) all belonging to the same user for networks other than Quiver.  So if I am getting access through you on both my laptop and desktop, using the remote instance id would result in each of the devices using a separate TOR path, while using userId would have them go through the same path.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

dborkan commented:
*[src/generic_core/globals.ts, line 46 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOLHtlP5VYXSUACGPCj:-KOLHrSusSWgwv0OHKuj:1451746356) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/globals.ts#L46)):*
> ```XML
>     enabled: false,
>     socksEndpoint: {address: '127.0.0.1', port: 9050},
>     auth: [socks.Auth.NOAUTH]
> ```

Also I'm curious why is this an array?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/generic_core/globals.ts, line 46 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6wM4TVvfHvUdhW9e:-KOLgoknyuAmjelXMXyi:311163463) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/globals.ts#L46)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I find this very strange.  What is being set globally by this?
</details>
It is a global setting of the reproxy settings. In order to initiate a socks connection, the auth must be negotiated. Before the only auth option that was negotiated was NOAUTH. Since now USERPASS is supported to some degree, I pulled the auth variable into the global settings to determine what type of auth to negotiate.

---

*[src/generic_core/globals.ts, line 46 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOLHtlP5VYXSUACGPCj:-KOLjM6D1IwRzisLiwe7:1641746265) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/globals.ts#L46)):*
<details><summary><i>Previously, dborkan wrote…</i></summary>
> Also I'm curious why is this an array?
</details>
It is an array because auth negotiation requires sending an array of auth options to the socks server, and then the socks server picks which option to continue auth with by sending a response. So a valid auth setting would be [USERPASS, NOAUTH] if userpass is preferred or [NOAUTH, USERPASS] if noauth is preferred or even [USERPASS] to indicate if userpass is not available to end connection.

---

*[src/generic_core/remote-connection.ts, line 157 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOLHEj40e8ca40brp3r:-KOLkCNQKcSiTEPZge7P:1170802582) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, dborkan wrote…</i></summary>
> This is kind of a nitpick given that multiple instance support may be going away with HotPot, but I think this should probably be the remote instance id, not the user id.  uProxy currently supports sharing with multiple devices (instances) all belonging to the same user for networks other than Quiver.  So if I am getting access through you on both my laptop and desktop, using the remote instance id would result in each of the devices using a separate TOR path, while using userId would have them go through the same path.
</details>
Okay, I can make that change. 

It is a bit unclear to me whether the Tor people would have an issue with that scenario. Where all a single user's devices use the same Tor path.

---

*[src/generic_ui/scripts/model.ts, line 78 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6k8pzaYvchpatRo7:-KOLh0FVP0IqQpm6zjT3:-1529577761) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_ui/scripts/model.ts#L78)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Will this `auth` parameter show up in Advanced Settings?
</details>
Yes it will, as part of the large JSON blob. Eventually I envision a general UI component to set reproxy settings by giving choices for auth. Right now, the only relevant reproxy settings are being set by the Tor UI component.

---

*[src/interfaces/uproxy_core_api.ts, line 94 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL7ASL-txZeD77HvuI:-KOLl7SUrNjjqm4Klwbd:-353343487) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/interfaces/uproxy_core_api.ts#L94)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I don't understand what this is for.
</details>
I put auth in the settings to allow users to choose which auth method they would like to use for reproxying. Right now, it is not strictly required since Tor is the only reproxy we are using, so it can be hard-coded in. Somewhat similar to how username and password are hard-coded in.

I would eventually like to see all these options in reproxySettings to abstract to an interface for any socks reproxy, instead of doing many Tor-specific choices that might be tougher to generalize later.

---

*[src/lib/socks/headers.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5x3PGGr63EZ1H8em:-KOLmCQGaNXzxPyQRb5v:1538599025) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L10)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Hmm.  I'm not sure how I feel about this.  There's something odd about mixing version numbers of two different versioned things in one enum.
</details>
I agree, thought about some other options that I didn't like. The alternative would be creating two things, a version enum (or just variable) for socks versions and then for socks userpass subnegotiation versions.

The Socks protocol is standard on using this format for versioning both itself and all its subprotocols.

---

*[src/lib/socks/headers.ts, line 239 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL4mBVQZmPd1jAUTIS:-KOLnFzlj29OjZJAChut:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L239)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: spaces around `*`
</details>
Done.

---

*[src/lib/socks/headers.ts, line 277 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6_z3b0DDNBi4yo1y:-KOLnaJOg2qqlCg0a2yr:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L277)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: two spaces between `;` and `//`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/lib/socks/headers.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5x3PGGr63EZ1H8em:-KOLtyoSoxdX6f4xmdMY:586518737) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L10)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> I agree, thought about some other options that I didn't like. The alternative would be creating two things, a version enum (or just variable) for socks versions and then for socks userpass subnegotiation versions.
> 
> The Socks protocol is standard on using this format for versioning both itself and all its subprotocols.
</details>
The Socks protocol *seems standard on using this format

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 157 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOLHEj40e8ca40brp3r:-KOLIH8EVk7FQpbOgzf3:1100653134) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Okay, I can make that change. 
> 
> It is a bit unclear to me whether the Tor people would have an issue with that scenario. Where all a single user's devices use the same Tor path.
</details>
Yeah.  I see @dborkan's logic, but I'm also fine with unifying all of a user's traffic ... especially since the situation is mostly hypothetical.

---

*[src/interfaces/uproxy_core_api.ts, line 94 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL7ASL-txZeD77HvuI:-KOLu-XAMWezDJE3Y241:2105424738) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/interfaces/uproxy_core_api.ts#L94)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> I put auth in the settings to allow users to choose which auth method they would like to use for reproxying. Right now, it is not strictly required since Tor is the only reproxy we are using, so it can be hard-coded in. Somewhat similar to how username and password are hard-coded in.
> 
> I would eventually like to see all these options in reproxySettings to abstract to an interface for any socks reproxy, instead of doing many Tor-specific choices that might be tougher to generalize later.
</details>
I assumed our behavior would be to always offer NOAUTH and USERPASS.  Is there a reason to make that configurable?

I might be able to imagine some alternative reproxying options for other proxies, but I'm not aware of a situation where we would want to let the user choose which auth modes to offer.

---

*[src/lib/socks/headers.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5x3PGGr63EZ1H8em:-KOLuMahUH9VE8pxoT0K:236363752) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L10)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> The Socks protocol *seems standard on using this format
</details>
OK.  I'd appreciate a bit more comment explaining that this is not just the SOCKS protocol version.

---

*[src/lib/socks/headers.ts, line 277 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6_z3b0DDNBi4yo1y:-KOLuYfS2pYFUYXcL69n:1373859459) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L277)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Done.
</details>
This is not the only place.  Try searching for `; //`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

trevj commented on src/generic_core/remote-connection.ts:
```diff
@@ -153,7 +153,8 @@ var generateProxyingSessionId_ = (): string => {
         allowNonUnicast: globals.settings.allowNonUnicast,
-        reproxy: globals.settings.reproxy
+        reproxy: globals.settings.reproxy,
+        userId: this.userId_
```
Do you need this? Seems like `RtcToNet` already knows the `userId` since its constructor accepts it as an argument.

trevj commented on src/generic_ui/scripts/model.ts:
```diff
@@ -72,7 +72,11 @@ export class Model {
+    reproxy: {
+      enabled: false,
+      socksEndpoint: {address: '127.0.0.1', port: 9050},
+      auth: [0]
```
Specifying the literals seems hard, even for advanced settings users! Did you consider having a boolean named something like `enableUserPass`?

trevj commented on src/lib/rtc-to-net/rtc-to-net.ts:
```diff
@@ -496,22 +496,45 @@ import ProxyConfig = require('./proxyconfig');
+              username: this.proxyConfig_.userId || 'user',
+              password: ''
+            };
+            var userpassRequest = socks_headers.composeUserPassRequest(userpass);
```
FYI, you can rewrite this as:
```typescript
const userpassRequest = socks_headers.composeUserPassRequest({
  username: this.proxyConfig_.userId || 'user'
  password: ''
};
```

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -65,6 +68,13 @@ export interface Request {
 
+// Represents a SOCKS username-password pair for USERPASS auth
+// @see interpretUserPassRequest
+export interface UserPass {
```
I think `UserPassRequest` would be a better name.

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
+//            +----+------+----------+------+----------+
+//            | 1  |  1   | 1 to 255 |  1   | 1 to 255 |
+//            +----+------+----------+------+----------+
+export function composeUserPassRequest(creds:UserPass) : ArrayBuffer {
```
I would suggest renaming `creds` -> `userPass`, what do you think?

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
+//            | 1  |  1   | 1 to 255 |  1   | 1 to 255 |
+//            +----+------+----------+------+----------+
+export function composeUserPassRequest(creds:UserPass) : ArrayBuffer {
+  var ulen :number = 2 * creds.username.length; // 2 bytes per char
```
No need for `:number` since the compiler can easily infer the type. Also, if something can be `const` it generally should.

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
+
+  requestBytes[0] = Version.VERSION1;  // Only support version 1 of UserPass protocol.
+  requestBytes[1] = ulen;
+  requestBytes.set(new Uint8Array(str2ab(creds.username)), 2);
```
```typescript
requestBytes.set(new Buffer(creds.username, 'ascii'), 2);
```

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
+  }
+
+  var ulen :number = requestBytes[1];
+  var username :string = ab2str(requestBytes.slice(2, 2 + ulen));
```
```typescript
const username = new Buffer(requestBytes.slice(2, 2 + ulen)).toString('ascii');
```

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
+}
+
+export function interpretUserPassResponse(buffer:ArrayBuffer) : boolean {
+  var responseBytes = new Uint8Array(buffer);
```
I think you should add a check for the length of `buffer`.

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
+
+export function interpretUserPassResponse(buffer:ArrayBuffer) : boolean {
+  var responseBytes = new Uint8Array(buffer);
+  return responseBytes[1] == 0 ? true : false;
```
No need for the ternary operator, this is equivalent:
```typescript
return responseBytes[1] === 0;
```
:-)

trevj commented on src/lib/socks/headers.ts:
```diff
@@ -484,3 +560,20 @@ export function interpretResponseBuffer(buffer:ArrayBuffer) : Response {
+// Helper methods for encoding strings to/from ArrayBuffer
+
+// ArrayBuffer to string
+function ab2str(buf:ArrayBuffer) : string {
```
We can use `Buffer` rather than (re-)adding these helper functions - I gave examples above where you've used these.

trevj replied:
BTW, we can use this API:
https://nodejs.org/dist/latest-v4.x/docs/api/buffer.html#buffer_buf_tostring_encoding_start_end

nirvantyagi commented:
Review status: 0 of 7 files reviewed at latest revision, 20 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 157 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxv-r2-157:-KOM7Qx_cufiM0NI5r1O:-147014756) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Do you need this? Seems like `RtcToNet` already knows the `userId` since its constructor accepts it as an argument.
</details>
Yes, RtcToNet has access to userId, but we need to pass it down into Session.

---

*[src/generic_ui/scripts/model.ts, line 78 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxw-r2-78:-KOM7fyMu_vnlRlxDKWQ:1854975180) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/generic_ui/scripts/model.ts#L78)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Specifying the literals seems hard, even for advanced settings users! Did you consider having a boolean named something like `enableUserPass`?
</details>
Yeah, I think for now we have to decided to remove the auth from advanced settings.

In general, I think the goal was to avoid having many different booleans describing different reproxy setups and instead have a separate UI component that would help the user navigate options and then automatically set the appropriate items.

---

*[src/lib/rtc-to-net/rtc-to-net.ts, line 521 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxz-r2-521:-KOM92M5HI352slYL0tc:-1043637872) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/rtc-to-net/rtc-to-net.ts#L521)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> FYI, you can rewrite this as:
> 
> ```typescript
> 
> const userpassRequest = socks_headers.composeUserPassRequest({
> 
>   username: this.proxyConfig_.userId || 'user'
> 
>   password: ''
> 
> };
> 
> ```
</details>
I wanted to log userpass, so I split it up. I wasn't having luck logging the userpassRequest since it is an ArrayBuffer. Would it be preferable to use arraybuffers.buf2str in order to log?

---

*[src/lib/socks/headers.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5x3PGGr63EZ1H8em:-KOM1w46xZ5kmTCAV8oy:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L10)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> OK.  I'd appreciate a bit more comment explaining that this is not just the SOCKS protocol version.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 275 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5C7tBDrisOF00LRo:-KON1RMziLGUfGUdyneS:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L275)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Please add unit tests for these four new functions to headers.spec.ts.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 277 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL6_z3b0DDNBi4yo1y:-KOMCnB_Wy2aO1jMKGz5:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L277)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This is not the only place.  Try searching for `; //`
</details>
Done.

---

*[src/lib/socks/headers.ts, line 73 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-73:-KOMA-uuU3gIg-_acX-r:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L73)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> I think `UserPassRequest` would be a better name.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 238 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-238:-KOMA1_EAb4odduw3go3:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L238)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> I would suggest renaming `creds` -> `userPass`, what do you think?
</details>
Done.

---

*[src/lib/socks/headers.ts, line 239 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-239:-KOMAwK_4tCH5E8D-yEO:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L239)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> No need for `:number` since the compiler can easily infer the type. Also, if something can be `const` it generally should.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 245 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-245:-KOMI8w6YpLnWhWrJpOV:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L245)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> ```typescript
> 
> requestBytes.set(new Buffer(creds.username, 'ascii'), 2);
> 
> ```
</details>
Done.

---

*[src/lib/socks/headers.ts, line 261 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-261:-KOMFcSTiCanjTGdZGBt:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L261)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> ```typescript
> 
> const username = new Buffer(requestBytes.slice(2, 2 + ulen)).toString('ascii');
> 
> ```
</details>
Done.

---

*[src/lib/socks/headers.ts, line 283 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-283:-KOMDYAl_oUKkrCmAB_1:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L283)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> I think you should add a check for the length of `buffer`.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 284 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-284:-KOMD1u5Cac37svix_is:-1998908372) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L284)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> No need for the ternary operator, this is equivalent:
> 
> ```typescript
> 
> return responseBytes[1] === 0;
> 
> ```
> 
> :-)
</details>
haha oops

---

*[src/lib/socks/headers.ts, line 567 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-567:-KOMErv3tvbRQKDpwsLW:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L567)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> BTW, we can use this API:
> 
> https://nodejs.org/dist/latest-v4.x/docs/api/buffer.html#buffer_buf_tostring_encoding_start_end
</details>
Done.

---

*[src/interfaces/uproxy_core_api.ts, line 94 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL7ASL-txZeD77HvuI:-KOM8XuhM2X8jvLkrfXg:-2119478400) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/interfaces/uproxy_core_api.ts#L94)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I assumed our behavior would be to always offer NOAUTH and USERPASS.  Is there a reason to make that configurable?
> 
> I might be able to imagine some alternative reproxying options for other proxies, but I'm not aware of a situation where we would want to let the user choose which auth modes to offer.
</details>
Chatting with Ben, I misunderstood how this worked. The array of auth options that are sent over is not in any ordered list. It is the socks server that has some ordering process to choose what auth method to use. In this case, we will not display options for auth and instead always send [USERPASS, NOAUTH], since these are the two auth methods that our sharer can currently support.

This is still a bit strange since if a server does choose USERPASS, we send a nonsense random username and password that will work for Tor, but doesn't actually authenticate anything. Eventually, we may like to add username and password as options for reproxy. It is okay for now, if we assume that servers that pick USERPASS (and then fail), would not support NOAUTH anyways (not the case for the Tor server).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

trevj commented:
Looking very good! Still a few more comments on the headers stuff but we're very close to ready AFAIC (that stuff is prone to error so important nail down early - thanks for adding tests!).

---

Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 157 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxv-r2-157:-KOQUN4zONXG7o-QtAWa:b-2b0vnm) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Yes, RtcToNet has access to userId, but we need to pass it down into Session.
</details>
I don't get it - the previous line passes this.userId_ to `RtcToNet`'s constructor.

---

*[src/lib/socks/headers.spec.ts, line 54 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQUk7pAc6kdclgAWBC:-KOQUk7qIcZ5agU-If17:b-xyrlww) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.spec.ts#L54)):*
> ```XML
> 
>   it('roundtrip userpass auth request', () => {
>     var userpass = {username: 'user', password: 'pass'}
> ```

```typescript
const userpass = {
  username: 'user',
  password: 'pass'
};
```

---

*[src/lib/socks/headers.ts, line 572 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOL5mGBCBp0Z7rFqOKh:-KOQVRHl9dcCNAtdlhiQ:ba1n0pc) ([raw file](https://github.com/uproxy/uproxy/blob/af598808fe7dce71c713513de4fd292a7b49aab9/src/lib/socks/headers.ts#L572)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I think these functions are already available from lib/arraybuffers/arraybuffers.ts.  You might as well just import that instead.
</details>
You're reminded me to continue removing that stuff :-)

---

*[src/lib/socks/headers.ts, line 239 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-239:-KOQUumC_GJKtozZlpLU:b-sp5wkk) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L239)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Done.
</details>
You're still explicitly declaring the type for `requestBytes`.

---

*[src/lib/socks/headers.ts, line 253 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQW9wRe1TfIVCbiQpJ:-KOQW9wRe1TfIVCbiQpK:b3q5yhy) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.ts#L253)):*
> ```XML
> 
> export function interpretUserPassRequest(buffer:ArrayBuffer) : UserPassRequest {
>   var requestBytes = new Uint8Array(buffer);
> ```

As for `interpretUserPassResponse`, I think you should add a check for the length of `buffer`.

---

*[src/lib/socks/headers.ts, line 261 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQVaZWWyFQo7A7ArBh:-KOQVaZWWyFQo7A7ArBi:b-welsg6) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.ts#L261)):*
> ```XML
>   }
>   const ulen = requestBytes[1];
>   const username = (new Buffer(requestBytes)).slice(2, 2 + ulen).toString('ascii');
> ```

So you are still assuming characters consume two bytes which isn't the case with ASCII. You should convert the strings to `Buffer`s and examine the size of the resulting `Buffer`s.

---

*[src/lib/socks/headers.ts, line 264 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQVnktDXvJO6QmIRbR:-KOQVnktDXvJO6QmIRbS:b-psmzu6) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.ts#L264)):*
> ```XML
>   const plen = requestBytes[2 + ulen];
>   const password = (new Buffer(requestBytes)).slice(2 + ulen + 1).toString('ascii');
>   return {username: username, password: password};
> ```

Formatting:
```typescript
return {
  username: username,
  password: password
};
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-:-KOQWR-zzEsgsHMf6H2N:bdcsdx1)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 7 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 157 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxv-r2-157:-KOQaiaxM7hxwpEP-cEb:b7n3xwe) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> I don't get it - the previous line passes this.userId_ to `RtcToNet`'s constructor.
</details>
Right. `userId` is passed into `RtcToNet`'s constructor which represents a single Getter-Sharer relationship. `RtcToNet` then creates and handles separate `Session` objects for each connection that the Getter is using. The `Session` object needs access to `userId` in order to complete socks auth, however it does not contain a backpointer to `RtcToNet`; it is only passed the `proxyConfig`.

---

*[src/lib/socks/headers.spec.ts, line 54 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQUk7pAc6kdclgAWBC:-KOQbxxMEa_6w1GaIF18:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.spec.ts#L54)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> ```typescript
> const userpass = {
>   username: 'user',
>   password: 'pass'
> };
> ```
</details>
Done.

---

*[src/lib/socks/headers.ts, line 239 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKy0-r2-239:-KOQdoUN3-FvklE91aYp:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/lib/socks/headers.ts#L239)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> You're still explicitly declaring the type for `requestBytes`.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 253 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQW9wRe1TfIVCbiQpJ:-KOQlTFWvpxgCz9Eyl7h:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.ts#L253)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> As for `interpretUserPassResponse`, I think you should add a check for the length of `buffer`.
</details>
Done.

---

*[src/lib/socks/headers.ts, line 261 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQVaZWWyFQo7A7ArBh:-KOQlW5zwRRQuCc2bcdX:bsfitz7) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.ts#L261)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> So you are still assuming characters consume two bytes which isn't the case with ASCII. You should convert the strings to `Buffer`s and examine the size of the resulting `Buffer`s.
</details>
No, I had switched all the logic away from 2 bytes/char (just removed a comment that said otherwise). This code slices correctly at index 2, since index 0 is version, index 1 is ulen.

---

*[src/lib/socks/headers.ts, line 264 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOQVnktDXvJO6QmIRbR:-KOQmOuk_sa4ahGDaRVY:b-896fix) ([raw file](https://github.com/uproxy/uproxy/blob/cce0d5a9d1df4eae22955f2fb30fc56a085902c8/src/lib/socks/headers.ts#L264)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Formatting:
> ```typescript
> return {
>   username: username,
>   password: password
> };
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

trevj commented:
Review status: 0 of 7 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 157 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxv-r2-157:-KOgTGGrOQAnIWXfzK2d:b-fgelhm) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, nirvantyagi (Nirvan Tyagi) wrote…</i></summary>
> Right. `userId` is passed into `RtcToNet`'s constructor which represents a single Getter-Sharer relationship. `RtcToNet` then creates and handles separate `Session` objects for each connection that the Getter is using. The `Session` object needs access to `userId` in order to complete socks auth, however it does not contain a backpointer to `RtcToNet`; it is only passed the `proxyConfig`.
</details>
But `RtcToNet` creates `Session`.

Don't have two places to set class-level fields: add `userId` to Session's constructor (alternative: add `userId` to `ProxyConfig` but remove it from `RtcToNet`'s constructor arg list).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

nirvantyagi commented:
Review status: 0 of 6 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/generic_core/remote-connection.ts, line 157 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2637#-KOHErQ4GpZuv-5vZKxv-r2-157:-KOjT1rKYYAKkiuG-SLO:b-5dqak1) ([raw file](https://github.com/uproxy/uproxy/blob/9c402646a2e8234fb4f8ce2c17c9b11ef08fde13/src/generic_core/remote-connection.ts#L157)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> But `RtcToNet` creates `Session`.
> 
> Don't have two places to set class-level fields: add `userId` to Session's constructor (alternative: add `userId` to `ProxyConfig` but remove it from `RtcToNet`'s constructor arg list).
</details>
Fixed! Removed `userId_` from `proxyConfig` and added it as optional parameter in `Session` constructor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2637)*
<!-- Sent from Reviewable.io -->

trevj commented:
Thank you! :+1:
